tests/test_smc.py

```diff
@@ -0,0 +1,164 @@
+    all_xbs = []
+    for _ in range(num_samples):
+        choice_idx = np.random.choice(np.arange(len(solvent_samples)))
+        solvent_x = solvent_samples[choice_idx].coords
+        ligand_x = enhanced.sample_from_log_weights(ligand_samples, ligand_log_weights, size=1)[0]
+        combined_x = np.concatenate([solvent_x[:-num_ligand_atoms], ligand_x], axis=0)
+        combined_box = solvent_samples[choice_idx].box
+        all_xbs.append((combined_x, combined_box))
+    return all_xbs
+
+
+def test_smc():
+    """
+    Generate samples from the equilibrium distribution at lambda=1
+    """
+
+    seed = 2021
+    np.random.seed(seed)
+
+    mol, torsion_idxs = test_ligands.get_biphenyl()
+
+    @jax.jit
+    def get_torsion(x_l):
+        ci = x_l[torsion_idxs[:, 0]]
+        cj = x_l[torsion_idxs[:, 1]]
+        ck = x_l[torsion_idxs[:, 2]]
+        cl = x_l[torsion_idxs[:, 3]]
+        # last [0] is used to return from a length-1 array
+        return bonded.signed_torsion_angle(ci, cj, ck, cl)[0]
+
+    ff = get_ff_am1cc()
+
+    ligand_masses = np.array([a.GetMass() for a in mol.GetAtoms()])
+    num_ligand_atoms = len(ligand_masses)
+
+    temperature = 300.0
+    pressure = 1.0
+
+    ubps, params, masses, coords, box = enhanced.get_solvent_phase_system(mol, ff)
+
+    cache_path = "test_smc_cache.pkl"
+    if not os.path.exists(cache_path):
+        print("Generating cache")
+
+        n_solvent_samples = 100  # should be maybe 1000
+        print(f"generating {n_solvent_samples} solvent samples")
+        solvent_xvbs = generate_solvent_samples(
+            coords, box, masses, ubps, params, temperature, pressure, seed, n_solvent_samples
+        )
+
+        n_ligand_batches = 1200  # should be 30k
+        print(f"generating ligand samples")
+        ligand_samples, ligand_log_weights = generate_ligand_samples(n_ligand_batches, mol, ff, temperature, seed)
+
+        with open(cache_path, "wb") as fh:
+            pickle.dump([solvent_xvbs, ligand_samples, ligand_log_weights], fh)
+
+    with open(cache_path, "rb") as fh:
+        print("Loading cache")
+        solvent_xvbs, ligand_samples, ligand_log_weights = pickle.load(fh)
+
+    n_endstate_samples = 5000
+    all_xbs = generate_endstate_samples(
+        n_endstate_samples, solvent_xvbs, ligand_samples, ligand_log_weights, num_ligand_atoms
+    )
+
+    # plot torsions at the end-states
+    end_state_torsions = []
+    for xb in all_xbs:
+        x = xb[0]
+        x_l = x[:-num_ligand_atoms]
+        end_state_torsions.append(get_torsion(x_l))
+
+    # plot histogram using asciiplotlib
+    fig = apl.figure()
+    fig.hist(*np.histogram(end_state_torsions, bins=25, range=(-np.pi, np.pi)))
+    fig.show()
+
+    bound_impls = []
+    for u, p in zip(ubps, params):
+        bound_impls.append(u.bind(p).bound_impl(np.float32))
+
+    # propagate with NPTMove
+    n_steps = 1000
+    npt_lamb = 1.0
+    seed = int(time.time())
```

Comment:
May want to freeze this to a constant in the test?
```suggestion
    seed = 2021
```

Reply:
Will freeze for actual test case - I need to think a bit more carefully about if we want each mover for a given lambda window to restart with the same seed.